File: docs-examples/example-java/src/test/java/io/micronaut/configuration/openapi/docs/SchemasTest.java
package io.micronaut.configuration.openapi.docs;

import io.swagger.v3.oas.models.media.Schema;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

class SchemasTest {

    @Test
    void schemaAnnotationCustomizesThePojo() {
        var schemas = OpenApiSpec.load().getComponents().getSchemas();

        Schema<?> pet = schemas.get("MyPet");
        assertNotNull(pet);
        assertEquals("Pet description", pet.getDescription());
        Schema<?> age = pet.getProperties().get("age");
        assertEquals("Pet age", age.getDescription());
        assertEquals(new BigDecimal("20"), age.getMaximum());
        Schema<?> name = pet.getProperties().get("name");
        assertEquals("Pet name", name.getDescription());
        assertEquals(20, name.getMaxLength());
        Schema<?> petType = schemas.get("PetType");
        assertEquals("string", petType.getType());
        assertEquals(List.of("DOG", "CAT"), petType.getEnum());
    }

    @Test
    void metaAnnotationAppliesTheSchema() {
        var operation = OpenApiSpec.load().getPaths().get("/pets").getPost();

        var schema = operation.getRequestBody().getContent().get("application/json").getSchema();
        assertEquals("#/components/schemas/MyPet", schema.get$ref());
    }

    @Test
    void genericsAreIncludedInTheSchemaName() {
        var openApi = OpenApiSpec.load();

        var response = openApi.getPaths().get("/").getPut().getResponses().get("200");
        assertEquals("#/components/schemas/Response_Pet_", response.getContent().get("application/json").getSchema().get$ref());
        Schema<?> schema = openApi.getComponents().getSchemas().get("Response_Pet_");
        assertEquals("#/components/schemas/MyPet", schema.getProperties().get("result").get$ref());
    }

    @Test
    void schemaNameCanBeChanged() {
        var openApi = OpenApiSpec.load();

        var response = openApi.getPaths().get("/named").getPut().getResponses().get("200");
        assertEquals("#/components/schemas/ResponseOfPet", response.getContent().get("application/json").getSchema().get$ref());
        assertTrue(openApi.getComponents().getSchemas().containsKey("ResponseOfPet"));
    }
}

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/configuration/openapi/docs/SchemasTest.kt
package io.micronaut.configuration.openapi.docs

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.math.BigDecimal

class SchemasTest {

    @Test
    fun schemaAnnotationCustomizesThePojo() {
        val schemas = OpenApiSpec.load().components.schemas

        val pet = schemas["MyPet"]
        assertNotNull(pet)
        assertEquals("Pet description", pet!!.description)
        val age = pet.properties["age"]!!
        assertEquals("Pet age", age.description)
        assertEquals(BigDecimal("20"), age.maximum)
        val name = pet.properties["name"]!!
        assertEquals("Pet name", name.description)
        assertEquals(20, name.maxLength)
        val petType = schemas["PetType"]!!
        assertEquals("string", petType.type)
        assertEquals(listOf("DOG", "CAT"), petType.enum)
    }

    @Test
    fun metaAnnotationAppliesTheSchema() {
        val operation = OpenApiSpec.load().paths["/pets"]!!.post

        val schema = operation.requestBody.content["application/json"]!!.schema
        assertEquals("#/components/schemas/MyPet", schema.`$ref`)
    }

    @Test
    fun genericsAreIncludedInTheSchemaName() {
        val openApi = OpenApiSpec.load()

        val response = openApi.paths["/"]!!.put.responses["200"]!!
        assertEquals("#/components/schemas/Response_Pet_", response.content["application/json"]!!.schema.`$ref`)
        val schema = openApi.components.schemas["Response_Pet_"]!!
        assertEquals("#/components/schemas/MyPet", schema.properties["result"]!!.`$ref`)
    }

    @Test
    fun schemaNameCanBeChanged() {
        val openApi = OpenApiSpec.load()

        val response = openApi.paths["/named"]!!.put.responses["200"]!!
        assertEquals("#/components/schemas/ResponseOfPet", response.content["application/json"]!!.schema.`$ref`)
        assertTrue(openApi.components.schemas.containsKey("ResponseOfPet"))
    }
}
